# Understanding `iloc` and `loc` in Pandas

When working with DataFrames, we often need to select specific rows and columns.

Pandas provides two important selection methods:

## 1. iloc

`iloc` stands for **Integer Location**.

It selects data based on **position numbers**.

Think of it as:

- First row → Position 0
- Second row → Position 1
- Third row → Position 2

---

## 2. loc

`loc` stands for **Location using Labels**.

It selects data based on **row labels (index values)**.

Think of it as:

- Student Roll Number
- Customer ID
- Transaction ID

Instead of position, we use labels.

---

Let's understand with examples.

In [3]:
import pandas as pd

df = pd.DataFrame({
    "Name": ["Venkatesh", "Rahul", "Priya", "Ankit"],
    "Age": [28, 25, 30, 27],
    "Salary": [100000, 80000, 120000, 90000]
})

df

,Name,Age,Salary
0,Venkatesh,28,100000
1,Rahul,25,80000
2,Priya,30,120000
3,Ankit,27,90000


## Selecting a Single Row using `iloc`

Syntax:

```python
df.iloc[row_position]

Here, 0 means:

First row
Position 0

Question:

What do you think will be returned?

In [ ]:
# Code Cell

df.iloc[0]

,0
Name,Venkatesh
Age,28
Salary,100000


## Selecting Multiple Rows using `iloc`

Python slicing rules apply.

Syntax:

```python
df.iloc[start:end]

In [ ]:
df.iloc[0:3]

,Name,Age,Salary
0,Venkatesh,28,100000
1,Rahul,25,80000
2,Priya,30,120000


returns:

Row 0
Row 1
Row 2

Row 3 is NOT included.

## Selecting Specific Rows and Columns using `iloc`

Syntax:

```python
df.iloc[row_range, column_range]

In [4]:
df.iloc[0:3, 0:2]

,Name,Age
0,Venkatesh,28
1,Rahul,25
2,Priya,30


Meaning:

Rows:

0
1
2

Columns:

0
1

Notice that the ending index is excluded for both rows and columns.

# Why Do We Need `loc`?

Right now:

```python
df.iloc[0]

In [ ]:
df.iloc[0]

,0
Name,Venkatesh
Age,28
Salary,100000


In [ ]:
df.loc[0]

,0
Name,Venkatesh
Age,28
Salary,100000


df.loc[0] appear to behave similarly.

This is because our index labels are:

0, 1, 2, 3

Let's change the index and see what happens.

In [5]:
df

,Name,Age,Salary
0,Venkatesh,28,100000
1,Rahul,25,80000
2,Priya,30,120000
3,Ankit,27,90000


In [6]:
df.index

RangeIndex(start=0, stop=4, step=1)

In [7]:

df2 = df.copy()

df2.index = [101, 102, 103, 104]

df2

,Name,Age,Salary
101,Venkatesh,28,100000
102,Rahul,25,80000
103,Priya,30,120000
104,Ankit,27,90000


## Understanding `iloc`

`iloc` always uses position.

Question:

What row will this return?

```python
df2.iloc[0]

In [ ]:
df2.iloc[0]

,101
Name,Venkatesh
Age,28
Salary,100000


In [8]:
df2.loc[0]

KeyError: 0

In [13]:
import pandas as pd

# 1. Define the live URL to the dataset repository
url = "https://raw.githubusercontent.com/dsrscientist/dataset1/master/titanic_train.csv"

# Production Guardrail: Network connectivity fail-safe for Google Colab
df = pd.read_csv(url)
print("✅ Live dataset successfully loaded from GitHub URL!")

# 2. Inspect the matrix layout dimensions
print(f"Dataset Shape (Rows, Columns): {df.shape}") # Output: (891, 12)

✅ Live dataset successfully loaded from GitHub URL!
Dataset Shape (Rows, Columns): (891, 12)


In [3]:
df.head(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


In [ ]:
df.iloc[:10, 2:5]

,Pclass,Name,Sex
0,3,"Braund, Mr. Owen Harris",male
1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female
2,3,"Heikkinen, Miss. Laina",female
3,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female
4,3,"Allen, Mr. William Henry",male
5,3,"Moran, Mr. James",male
6,1,"McCarthy, Mr. Timothy J",male
7,3,"Palsson, Master. Gosta Leonard",male
8,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female
9,2,"Nasser, Mrs. Nicholas (Adele Achem)",female


In [ ]:
df.loc[:10, 2:5]

TypeError: cannot do slice indexing on Index with these indexers [2] of type int

In [15]:
df.head()

,Name,Age,Salary
0,Venkatesh,28,100000
1,Rahul,25,80000
2,Priya,30,120000
3,Ankit,27,90000


In [ ]:
df.iloc[51:75, 2:5]

,Pclass,Name,Sex
51,3,"Nosworthy, Mr. Richard Cater",male
52,1,"Harper, Mrs. Henry Sleeper (Myna Haxtun)",female
53,2,"Faunthorpe, Mrs. Lizzie (Elizabeth Anne Wilkin...",female
54,1,"Ostby, Mr. Engelhart Cornelius",male
55,1,"Woolner, Mr. Hugh",male
56,2,"Rugg, Miss. Emily",female
57,3,"Novel, Mr. Mansouer",male
58,2,"West, Miss. Constance Mirium",female
59,3,"Goodwin, Master. William Frederick",male
60,3,"Sirayanian, Mr. Orsen",male


## Understanding `loc`

`loc` uses labels.

Question:

What row will this return?

```python
df2.loc[101]

In [14]:
df.loc[73]

KeyError: 73

In [ ]:
df2.loc[101]

,101
Name,Venkatesh
Age,28
Salary,100000


Remember:

loc uses labels.

Do we have a label called 0?

No.

Therefore this will produce an error.

# Selecting Rows and Columns using `loc`

Syntax:

```python
df.loc[row_label, column_label]

In [ ]:
df.loc[0, "Age"]

np.int64(28)

# Important Difference in Slicing

## loc

Ending value is INCLUDED.

Example:

```python
df.loc[0:2]

In [ ]:
df.loc[0:2]

,Name,Age,Salary
0,Venkatesh,28,100000
1,Rahul,25,80000
2,Priya,30,120000


In [ ]:
df.loc[100:150, ["Pclass", "Age", "Survived"]]

,Pclass,Age,Survived
100,3,28.0,0
101,3,NaN,0
102,1,21.0,0
103,3,33.0,0
104,3,37.0,0
105,3,28.0,0
106,3,21.0,1
107,3,NaN,1
108,3,38.0,0
109,3,NaN,1


## iloc

Ending value is EXCLUDED.

Example:

```python
df.iloc[0:2]

In [ ]:
df.iloc[0:2]

,Name,Age,Salary
0,Venkatesh,28,100000
1,Rahul,25,80000


# Industry Example

Suppose we have customer data.

Customer IDs:

101
102
103
104

If we want customer 103:

```python
df.loc[103]

# Industry Example

Suppose we have customer data.

Customer IDs:

101
102
103
104

If we want customer 103:

```python
df.loc[103]

This is why loc is commonly used with:

Customer IDs
Order IDs
Transaction IDs
Dates

If we simply want:

First 100 rows
Last 50 rows
First 10 columns

we usually use:

df.iloc

because positions are easier to work with.

In [43]:
df.head(10)

,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
PassengerId,,,,,,,,,,,
1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S


In [44]:
df = df.set_index("PassengerId")

df.head()

KeyError: "None of ['PassengerId'] are in the columns"

In [ ]:
df.loc[1]

Survived                          0
Pclass                            3
Name        Braund, Mr. Owen Harris
Sex                            male
Age                            22.0
SibSp                             1
Parch                             0
Ticket                    A/5 21171
Fare                           7.25
Cabin                           NaN
Embarked                          S
Name: 1, dtype: object

In [ ]:
df.iloc[0]

Survived                          0
Pclass                            3
Name        Braund, Mr. Owen Harris
Sex                            male
Age                            22.0
SibSp                             1
Parch                             0
Ticket                    A/5 21171
Fare                           7.25
Cabin                           NaN
Embarked                          S
Name: 1, dtype: object

In [ ]:
df.iloc[0]

Survived                          0
Pclass                            3
Name        Braund, Mr. Owen Harris
Sex                            male
Age                            22.0
SibSp                             1
Parch                             0
Ticket                    A/5 21171
Fare                           7.25
Cabin                           NaN
Embarked                          S
Name: 1, dtype: object

In [ ]:
mask = (df["Age"] < 18 ) & (df["Sex"] == "male")

print(type(mask))
mask.values

<class 'pandas.core.series.Series'>


array([False, False, False, False, False, False, False,  True, False,
       False, False, False, False, False, False, False,  True, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False,  True, False, False, False,
       False, False, False, False, False,  True, False, False, False,
        True, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False,  True, False, False,
       False, False, False, False, False,  True, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False,  True,
       False, False,

In [ ]:
df.iloc[mask.values]

,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
PassengerId,,,,,,,,,,,
8,0,3,"Palsson, Master. Gosta Leonard",male,2.00,3,1,349909,21.0750,NaN,S
17,0,3,"Rice, Master. Eugene",male,2.00,4,1,382652,29.1250,NaN,Q
51,0,3,"Panula, Master. Juha Niilo",male,7.00,4,1,3101295,39.6875,NaN,S
60,0,3,"Goodwin, Master. William Frederick",male,11.00,5,2,CA 2144,46.9000,NaN,S
64,0,3,"Skoog, Master. Harald",male,4.00,3,2,347088,27.9000,NaN,S
79,1,2,"Caldwell, Master. Alden Gates",male,0.83,0,2,248738,29.0000,NaN,S
87,0,3,"Ford, Mr. William Neal",male,16.00,1,3,W./C. 6608,34.3750,NaN,S
126,1,3,"Nicola-Yarred, Master. Elias",male,12.00,1,0,2651,11.2417,NaN,C
139,0,3,"Osen, Mr. Olaf Elon",male,16.00,0,0,7534,9.2167,NaN,S


In [ ]:
df.loc[(df["Age"] < 18 ) & (df["Sex"] == "male")]

,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
PassengerId,,,,,,,,,,,
8,0,3,"Palsson, Master. Gosta Leonard",male,2.00,3,1,349909,21.0750,NaN,S
17,0,3,"Rice, Master. Eugene",male,2.00,4,1,382652,29.1250,NaN,Q
51,0,3,"Panula, Master. Juha Niilo",male,7.00,4,1,3101295,39.6875,NaN,S
60,0,3,"Goodwin, Master. William Frederick",male,11.00,5,2,CA 2144,46.9000,NaN,S
64,0,3,"Skoog, Master. Harald",male,4.00,3,2,347088,27.9000,NaN,S
79,1,2,"Caldwell, Master. Alden Gates",male,0.83,0,2,248738,29.0000,NaN,S
87,0,3,"Ford, Mr. William Neal",male,16.00,1,3,W./C. 6608,34.3750,NaN,S
126,1,3,"Nicola-Yarred, Master. Elias",male,12.00,1,0,2651,11.2417,NaN,C
139,0,3,"Osen, Mr. Olaf Elon",male,16.00,0,0,7534,9.2167,NaN,S


In [ ]:
df.iloc

In [55]:
import pandas as pd
import numpy as np

orders = pd.DataFrame({
    "Order_ID": [1001, 1002, 1002, 1004, 1005],
    "Customer_Name": ["Rahul", "Priya", "Priya", "Vijay", None],
    "Age": [25, np.nan, np.nan, "twenty-eight", 32],
    "Order_Value": [500, 1200, 1200, np.nan, 800],
    "Order_Date": [
        "01-03-2025",
        "01-03-2025",
        "01-03-2025",
        "15/03/2025",
        "2025-03-20"
    ]
})

orders

,Order_ID,Customer_Name,Age,Order_Value,Order_Date
0,1001,Rahul,25,500.0,01-03-2025
1,1002,Priya,NaN,1200.0,01-03-2025
2,1002,Priya,NaN,1200.0,01-03-2025
3,1004,Vijay,twenty-eight,NaN,15/03/2025
4,1005,None,32,800.0,2025-03-20


In [ ]:
# inspect the data

orders.shape

(5, 5)

In [ ]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Order_ID       5 non-null      int64  
 1   Customer_Name  4 non-null      object 
 2   Age            3 non-null      object 
 3   Order_Value    4 non-null      float64
 4   Order_Date     5 non-null      object 
dtypes: float64(1), int64(1), object(3)
memory usage: 332.0+ bytes


In [ ]:
print(orders.isnull())

orders.isnull().sum()



   Order_ID  Customer_Name    Age  Order_Value  Order_Date
0     False          False  False        False       False
1     False          False   True        False       False
2     False          False   True        False       False
3     False          False  False         True       False
4     False           True  False        False       False


0    False
1    False
2     True
3    False
4    False
dtype: bool

Swiggy wants:

1. Average customer age
2. Total revenue
3. Total number of orders

Can we trust these numbers?

## **Cleaning Checklist:**

Every new dataset:

Step 1
- Inspect
   * head()
   * shape
   * info()

↓

Step 2
- Missing Values
   * isnull()

↓

Step 3
- Duplicates
   * duplicated()

↓

Step 4
- Data Types
   * info()
   * to_numeric()
   * to_datetime()

↓

Step 5
- Validation
   * isnull().sum()
   * duplicated().sum()
   * info()

↓

Step 6
Only then do analysis

# Explore & Audit a Dataset

Before cleaning data, we must first understand what problems exist.

Think like a Data Analyst.

Imagine your manager gives you a CSV file and asks:

"Can we use this data for reporting or machine learning?"

Never start cleaning immediately.

First, audit the dataset.

The workflow is:

AUDIT → DECIDE → FIX → VALIDATE

Today we will learn the AUDIT phase.

We will use the Titanic dataset because:

- It contains missing values
- It contains numeric and categorical columns
- It is widely used in Data Science
- We can focus on cleaning instead of understanding the business domain

# Audit Step 1: Check Dataset Size

Question:

How much data do we have?

The shape attribute returns:

(Number of Rows, Number of Columns)

Why is this important?

- More rows generally means more information.
- More columns means more features.
- Helps estimate the size of the cleaning task.

In [ ]:
import pandas as pd

# 1. Define the live URL to the dataset repository
url = "https://raw.githubusercontent.com/dsrscientist/dataset1/master/titanic_train.csv"

# Production Guardrail: Network connectivity fail-safe for Google Colab
df = pd.read_csv(url)
print("✅ Live dataset successfully loaded from GitHub URL!")

# 2. Inspect the matrix layout dimensions
print(f"Dataset Shape (Rows, Columns): {df.shape}") # Output: (891, 12)

✅ Live dataset successfully loaded from GitHub URL!
Dataset Shape (Rows, Columns): (891, 12)


In [ ]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


Titanic Dataset Shape:

(891, 12)

Interpretation:

891 passengers
12 columns/features

Question to students:

Would you manually inspect 891 rows one by one?

Answer: No.

This is why Pandas exists.

# Audit Step 2: Check Column Names and Data Types

Question:

What information is available?

What type of information is stored?

Examples:

- Age        -> Numeric
- Name       -> Text
- Sex        -> Category
- Fare       -> Numeric

Data types help us understand:

- Which columns can be used for mathematical operations
- Which columns are categories
- Which columns may need cleaning

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Common Data Types:

int64
    Whole numbers
    Example: PassengerId, Survived

float64
    Decimal numbers
    Example: Age, Fare

object
    Text / Strings
    Example: Name, Sex, Ticket

Question:

Can we calculate average Age?

Yes

Can we calculate average Name?

No

Data type determines what operations are possible.

# Feature Descriptions

| Column | Description | Example |
|----------|----------|----------|
| PassengerId | Unique identifier for each passenger | 1, 2, 3 |
| Survived | Whether passenger survived | 0 = No, 1 = Yes |
| Pclass | Passenger ticket class | 1 = First Class, 2 = Second Class, 3 = Third Class |
| Name | Passenger name | Allen, Mr. William |
| Sex | Gender | male, female |
| Age | Passenger age in years | 22, 38 |
| SibSp | Number of siblings/spouses travelling together | 0, 1, 2 |
| Parch | Number of parents/children travelling together | 0, 1, 2 |
| Ticket | Ticket number | A/5 21171 |
| Fare | Ticket price paid | 7.25, 71.28 |
| Cabin | Cabin number assigned on ship | C85, B57 |
| Embarked | Port where passenger boarded the ship | S, C, Q |

# Audit Step 3: Find Missing Values

One of the most common problems in real-world datasets is missing data.

Missing values are represented as:

NaN = Not a Number

Question:

How many values are missing in each column?

Let's investigate.

In [ ]:
df.isnull().sum()

,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,177
SibSp,0
Parch,0
Ticket,0
Fare,0


Look carefully at the output.

Questions:

Which column has the most missing values?

Which column looks almost complete?

Can we trust every column equally?

Notice:

Age has many missing values.

Cabin has a huge number of missing values.

# Missing Percentage

Counts are useful.

Percentages are even more useful.

Question:

Age has 177 missing values.

Is that bad?

We need context.

177 out of 891 is different from:

177 out of 10,000

Let's calculate percentages.

In [56]:
missing_percent = df.isnull().mean() * 100

missing_percent.round(2)

Survived     0.00
Pclass       0.00
Name         0.00
Sex          0.00
Age         19.87
SibSp        0.00
Parch        0.00
Ticket       0.00
Fare         0.00
Cabin       77.10
Embarked     0.22
dtype: float64


Age has around 20% missing values.

Would you:

1. Drop the entire Age column?
2. Fill the missing values?
3. Leave it unchanged?

Don't answer yet.

We will learn how to decide later.

# Audit Step 4: Check for Duplicate Rows

Duplicate records can cause incorrect statistics.

Example:

A customer order appears twice.

Revenue becomes double-counted.

Question:

How many completely duplicated rows exist?

In [ ]:
df.duplicated().sum()

np.int64(0)

# Audit Step 5: Understand Important Columns

Before cleaning, we should understand the distribution of important columns.

Let's inspect:

- Sex
- Passenger Class (Pclass)
- Age

In [57]:
print("Sex Distribution")
print(df["Sex"].value_counts())

print("\n" + "-"*50)

print("Passenger Class Distribution")
print(df["Pclass"].value_counts())

Sex Distribution
Sex
male      577
female    314
Name: count, dtype: int64

--------------------------------------------------
Passenger Class Distribution
Pclass
3    491
1    216
2    184
Name: count, dtype: int64


# Concept: Missing Values

Imagine you are working at PayPal.

Your manager asks:

"Can we build a model to predict whether a customer will churn?"

You open the dataset and see:

Age = Missing

Income = Missing

Country = Missing

Question:

Should we remove these records?

Should we guess the missing values?

Should we remove the entire column?

This is one of the most important decisions in Data Science.

There is no single answer.

The correct strategy depends on:

1. How much data is missing?
2. How important is the column?
3. Is the column numeric or categorical?

# The Three Common Strategies

When we find missing values, we usually choose one of three approaches.

Strategy 1: Drop

Remove rows or columns.

Strategy 2: Fill with a Constant

Examples:

Age = 0

City = "Unknown"

Strategy 3: Impute

Estimate a reasonable value using existing data.

Examples:

Mean

Median

Mode

# Important Beginner Confusion

Many students confuse:

dropna()

and

drop(columns=[...])

These are completely different.

dropna()

    Removes ROWS

drop(columns=[...])

    Removes COLUMNS

Notice:

Cabin is missing for most passengers.

Question:

If 77% of Cabin values are missing,
is it worth keeping?

Usually no.

This is why many Titanic analyses remove Cabin.